### ISSUE TITLE
Intermittent Spawn subprocess (fork()) "setup environment" error.

### ISSUE THREAD

**Issue by remomueller** [phusion/passenger#2092]
> ## Issue report
> Fill in as much as possible so that we can understand, find and fix the problem.
> **Are you sure this is a bug in Passenger?**
Please read our troubleshooting and debugging guides first:
https://www.phusionpassenger.com/library/admin/troubleshooting/
https://www.phusionpassenger.com/library/admin/

Yes, this occurs in 5.3.0, 5.3.1, and 5.3.2. This error does not occur in 5.2.3.

> **Please try with the newest version of Passenger to avoid issues that have already been fixed**
> **Question 1: What is the problem?**
> - What is the expected behavior?
> - What is the actual behavior?
> - How can we reproduce it? Please try to provide a sample application (or Virtual Machine) demonstrating the issue. Otherwise, if we can't reproduce it, we might have to ask you a number of followup questions or run certain commands to try and figure out the problem
> Be as detailed as possible in your descriptions, include any logs and stack traces (don't just cut/paste the error, provide some logging before that too).
> (if you are requesting a feature instead of reporting an issue, describe here what you have in mind and how it would help you)

The application will frequently create an error page with the following:

Cannot lookup up system user database entry for user 'remo': Undefined error: 0 (errno=0)

![screen shot 2018-06-13 at 9 43 14 am](https://user-images.githubusercontent.com/316479/41355266-1917c7c4-6eef-11e8-8c5e-6c2c760549a9.png)

```
PID
98341
Backtrace
     in 'bool Passenger::SpawningKit::HandshakePerform::checkCurrentState()' (Perform.h:237)
     in 'void Passenger::SpawningKit::HandshakePerform::waitUntilSpawningFinished(boost::unique_lock<boost::mutex> &)' (Perform.h:212)
     in 'Passenger::SpawningKit::Result Passenger::SpawningKit::HandshakePerform::execute()' (Perform.h:1740)
     in 'void Passenger::SpawningKit::SmartSpawner::internalStartPreloader(Passenger::SpawningKit::Config &, Passenger::SpawningKit::HandshakeSession &, const Json::Value &)' (SmartSpawner.h:421)
     in 'void Passenger::SpawningKit::SmartSpawner::startPreloader()' (SmartSpawner.h:310)
     in 'virtual Passenger::SpawningKit::Result Passenger::SpawningKit::SmartSpawner::spawn(const Passenger::SpawningKit::AppPoolOptions &)' (SmartSpawner.h:1281)
     in 'void Passenger::ApplicationPool2::Group::spawnThreadRealMain(const SpawningKit::SpawnerPtr &, const Passenger::ApplicationPool2::Options &, unsigned int)' (SpawningAndRestarting.cpp:95)
User and group
uid=0(root) gid=0(wheel) groups=0(wheel),1(daemon),2(kmem),3(sys),4(tty),5(operator),8(procview),9(procmod),20(staff),29(certusers),80(admin)
Ulimits
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
file size               (blocks, -f) unlimited
max locked memory       (kbytes, -l) unlimited
max memory size         (kbytes, -m) unlimited
open files                      (-n) 256
pipe size            (512 bytes, -p) 1
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 1064
virtual memory          (kbytes, -v) unlimited
```

> **Question 2: Passenger version and integration mode**:
> - For example: open source 5.0.26 standalone; enterprise 5.0.21/nginx

This is passenger (5.3.0, 5.3.1), 5.3.2 compiled from source with nginx 1.14.0: [MacOS installation](https://github.com/remomueller/documentation/tree/master/macos). It also happens on CentOS environment.

> **Question 3: OS or Linux distro, platform (including version)**:
> - For example: Debian 8, x86_64 or OS X 10.10 Yosemite, x86_64

MacOS 10.13.5 (Also happens on CentOS Linux release 7.3.1611)

> **Question 4: Passenger installation method**:

[ ] RubyGems + Gemfile
[ ] RubyGems, no Gemfile
[ ] Phusion APT repo
[ ] Phusion YUM repo
[ ] OS X Homebrew
[ ] source tarball
[X] Other, please specify:

Compiled from source: [MacOS installation](https://github.com/remomueller/documentation/tree/master/macos)

> **Question 5: Your app's programming language (including any version managers) and framework (including versions)**:
> - For example: Ruby 2.3.0, RVM, Rails 5.0.0; Node.js 4.0.0 with Express 4.13.4

RVM 1.29.3, Ruby 2.5.1, Rails 5.2.0, Node.js v6.14.12

> **Question 6: Are you using a PaaS and/or containerization? If so which one?**
> - For example: Heroku, Amazon Container Services, Docker 1.9 with an image based on passenger-docker

None. (development and staging environment)

> **Question 7: Anything else about your setup that we should know?**

I ran into this with 5.3.0 initially, and have tried both 5.3.1 and 5.3.2 as they've been released. The issue usually shows up a day after initially installing, at which point it shows up pretty consistently. I also run multiple passenger apps as well. I've also tried isolating and running nginx 1.14.0 on 5.2.3 which worked and did not encounter the issue.

**Comment by remomueller** [comment]
Also, not I didn't include this above, but it also has this more detailed error after "Cannot lookup up system user ..."

```
PassengerAgent(51819,0x7fff9f1ad380) malloc: *** mach_vm_map(size=1048576) failed (error code=268435459)
PassengerAgent(51819,0x7fff9f1ad380) malloc: *** error: can't allocate region securely
PassengerAgent(51819,0x7fff9f1ad380) malloc: *** set a breakpoint in malloc_error_break to debug
Error: Cannot lookup up system user database entry for user 'remo': Undefined error: 0 (errno=0)
```

**Comment by CamJN** [comment]
Unfortunately the `mach`/`malloc` stuff is a red herring. It's strange that lookup of a normal user account is failing. Is this mac bound to an active directory service or similar?

**Comment by remomueller** [comment]
Not that I'm aware (in terms of being part of active directory lookup). The account is a local account, no remote authentication.

**Comment by CamJN** [comment]
Here's a simplified test program, that you can compile with `clang++ test.cpp`, and run like `./a.out remo`, can you run it for a while and see if it prints anything? 

```
#include <iostream>
#include <pwd.h>
#include <string>
#include <unistd.h>

using namespace std;

int main(int argc, const char** argv) {
  string username = argv[1];
  struct passwd pwd, *userInfo;
  long pwdBufSize;
  char *pwdBuf;

  pwdBufSize = std::max<long>(1024 * 128, sysconf(_SC_GETPW_R_SIZE_MAX));
  pwdBuf = new char[pwdBufSize];

  int ret = 0;
  int count = 0;
  do {
    ret = getpwnam_r(username.c_str(), &pwd, pwdBuf, pwdBufSize, &userInfo);
    count++;
  } while (ret == 0);
  cout << "Failed lookup of system user database entry for user '" << username << "' , after " << count << " tries, errno = " << ret << endl;
  return 0;
}
```

**Comment by remomueller** [comment]
I ran it, and it didn't provide any output, just ran the loop until I hit Ctrl-C.

**Comment by remomueller** [comment]
Any updates on this? I've reverted back to v5.2.3 for the time being due to the instability in the 5.3.x series.

**Comment by CamJN** [comment]
No, I've been unable to reproduce the issue, and I run Passenger all day every day on my mac. Until we can figure out why that call is failing there isn't anything we can do about it.

**Comment by remomueller** [comment]
Ok, I'll keep trying the new releases to let you know if anything changes, thanks.
